### PR TITLE
.Net: Add Mssql Server connector, using vector search with plain tables

### PR DESCRIPTION
**Pull request by kbeaugrand**
### Motivation and Context

Adding MSSQL Server connector.

### Description

SQL server connector works when vector search capability is enabled in SQL server (In azure  it works), but in on prem scenarios, vector search is not accessible. 
This PR uses plain tables to provide vector search using SQL features.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

### CODE REVIEW

**Comment by yorek** [comment]
Since we also have a connector using the native vector functions (here: https://github.com/microsoft/semantic-kernel/tree/main/dotnet/src/Connectors/Connectors.Memory.SqlServer) can we call this connector `Microsoft.SemanticKernel.Connectors.SqlServer.Classic` instead of `Microsoft.SemanticKernel.Connectors.MssqlServer`.

That would help to bring clarity and have a clear differentiation between the one supporting native vectors functions and the one using a "classic" approach.

**Comment by kbeaugrand** [comment]
@RogerBarreto and @yorek as discussed internally, I renamed the connector to "classic" to match the clarity objectives regarding native approach...

Thank you!

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+                    CREATE TABLE {this.GetFullTableName(this._configuration.MemoryTableName)}
+                    (   [id] UNIQUEIDENTIFIER NOT NULL,
+                        [key] NVARCHAR(256)  NOT NULL,
+                        [collection] NVARCHAR(256) NOT NULL,
+                        [metadata] TEXT,
+                        [embedding] TEXT,
+                        [timestamp] DATETIMEOFFSET,
+                        PRIMARY KEY ([id]),
+                        FOREIGN KEY ([collection]) REFERENCES {this.GetFullTableName(this._configuration.MemoryCollectionTableName)}([id]) ON DELETE CASCADE,
+                        CONSTRAINT UK_{this._configuration.MemoryTableName} UNIQUE([collection], [key])
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:63]: Unless I missed something, there's mostly no escaping for these identifiers (in general, not only this constraint). This should be added.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    /// <inheritdoc />
+    public async IAsyncEnumerable<string> GetCollectionsAsync([EnumeratorCancellation] CancellationToken cancellationToken = default)
+    {
+        using (SqlCommand command = this._connection.CreateCommand())
+        {
+            command.CommandText = $"SELECT [id] FROM {this.GetFullTableName(this._configuration.MemoryCollectionTableName)}";
+
+            using var dataReader = await command.ExecuteReaderAsync(cancellationToken).ConfigureAwait(false);
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:137]: nit: Sometimes `var` is used, sometimes explicit type. Should be consistent.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+            queryColumns += ", [embedding]";
+        }
+
+        using SqlCommand entryCommand = this._connection.CreateCommand();
+
+        entryCommand.CommandText = $@"
+                                    SELECT {queryColumns}
+                                    FROM {this.GetFullTableName(this._configuration.MemoryTableName)}
+                                    WHERE [collection] = @collection
+                                    AND [key]=@key";
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:188]: nit: Spaces around `=`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+    }
+
+    /// <inheritdoc />
+    public async Task DeleteAsync(string collectionName, string key, CancellationToken cancellationToken = default)
+    {
+        collectionName = NormalizeIndexName(collectionName);
+
+        using SqlCommand cmd = this._connection.CreateCommand();
+
+        cmd.CommandText = $"DELETE FROM {this.GetFullTableName(this._configuration.MemoryTableName)} WHERE [collection] = @collectionName AND [key]=@key";
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:252]: nit: Spaces around `=`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+        collectionName = NormalizeIndexName(collectionName);
+
+        using SqlCommand cmd = this._connection.CreateCommand();
+
+        cmd.CommandText = $@"
+                MERGE INTO {this.GetFullTableName(this._configuration.MemoryTableName)}
+                USING (SELECT @key) as [src]([key])
+                ON {this.GetFullTableName(this._configuration.MemoryTableName)}.[key] = [src].[key]
+                WHEN MATCHED THEN
+                    UPDATE SET metadata=@metadata, embedding=@embedding, timestamp=@timestamp
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:386]: nit: Spaces around `=`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+        cmd.Parameters.AddWithValue("@metadata", metadata ?? (object)DBNull.Value);
+        cmd.Parameters.AddWithValue("@embedding", embedding);
+        cmd.Parameters.AddWithValue("@timestamp", timestamp ?? (object)DBNull.Value);
+
+        await cmd.ExecuteNonQueryAsync(cancellationToken).ConfigureAwait(false);
+    }
+
+    private string GetFullTableName(string tableName)
+    {
+        return $"[{this._configuration.Schema}].[{tableName}]";
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:425]: Proper escaping. I.e. what if the `tableName` is `foo]bar` (yes, that's stupid, but still).

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+
+        if (withEmbedding)
+        {
+            entry.Embedding = new ReadOnlyMemory<float>(JsonSerializer.Deserialize<IEnumerable<float>>(dataReader.GetString(dataReader.GetOrdinal("embedding")))!.ToArray());
+        }
+
+        return entry;
+    }
+
+    // Note: "_" is allowed in Postgres, but we normalize it to "-" for consistency with other DBs
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:454]: Postgres? ;)

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+    private static string NormalizeIndexName(string index)
+    {
+        index = s_replaceIndexNameCharsRegex.Replace(index.Trim().ToLowerInvariant(), ValidSeparator);
+
+        return index;
+    }
+
+    public void Dispose()
+    {
+        if (this._connection != null)
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:467]: I don't have a strong opinion about this, but calling `this._connection.Dispose();` is enough. Setting to `null` and doing the checks does not add any value (calling `Dispose` multiple times is fine).

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        }
+
+        if (limit <= 0)
+        {
+            yield break;
+        }
+
+        IAsyncEnumerable<(SqlServerMemoryEntry, double)> results = this._dbClient.GetNearestMatchesAsync(
+            collectionName: collectionName,
+            embedding: JsonSerializer.Serialize(embedding.ToArray()),
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs:142]: `ToArray` is not needed (wastes extra allocation). `JsonSerializer` can serialize `ROM<float>`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+            yield return await this.InternalUpsertAsync(collectionName, record, cancellationToken).ConfigureAwait(false);
+        }
+    }
+
+    /// <inheritdoc />
+    private MemoryRecord GetMemoryRecordFromEntry(SqlServerMemoryEntry entry)
+    {
+        return MemoryRecord.FromJsonMetadata(
+            json: entry.MetadataString,
+            embedding: entry.Embedding ?? new ReadOnlyMemory<float>(),
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs:209]: Use `ReadOnlyMemory<float>.Empty`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs [hunk 11]
```diff
@@ -0,0 +1,240 @@
...
+    /// <inheritdoc />
+    private async Task<string> InternalUpsertAsync(string collectionName, MemoryRecord record, CancellationToken cancellationToken)
+    {
+        record.Key = record.Metadata.Id;
+
+        await this._dbClient.UpsertAsync(
+            collectionName: collectionName,
+            key: record.Key,
+            metadata: record.GetSerializedMetadata(),
+            embedding: JsonSerializer.Serialize(record.Embedding.ToArray()),
```
Comment (cincuranet) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs:224]: Same as above about `ToArray`.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs (hunk 1 above)
Comment (kbeaugrand) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: Added escaped characters.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+    {
+        this._connectionString = connectionString;
+        this._configuration = configuration;
+
+        this._connection = new SqlConnection(this._connectionString);
+        this._connection.Open();
+    }
+
+    /// <inheritdoc />
+    public async Task CreateTablesAsync(CancellationToken cancellationToken)
```
[the remark this answers is not in the document]
Comment (kbeaugrand) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: Unfortunately, this is needed.

I think I've got 3 possibilities: 
1. Implement a synchronous method to create needed tables at startup (during the constructor) -> risk to have a long running job during constructor which is not fine
2. Implement the builder extension as this: 
    ```
    return builder.WithMemoryStore<SqlServerMemoryStore>(_ => SqlServerMemoryStore.ConnectAsync(connectionString, config)
                                                                               .ConfigureAwait(false)
                                                                               .GetAwaiter()
                                                                               .GetResult());
                                                                               
    ```  
    This will block the thread during the instanciation of the memoryDb, still not fine I think.
3. Let the user providing and instancing the memory store himself and provide the instance in the builder using the ``public MemoryBuilder WithMemoryStore(IMemoryStore store)`` method

I personally think the .3 is more relevant. More similar to the SQLite implementation. 
@westey-m what is you opinion?

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs (hunk 6 above)
Comment (kbeaugrand) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: Added escaping some characters in SLQL object names.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (cincuranet) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: IMHO both 1. and 2. should be no-go.

[the remark this answers is not in the document]
Comment (westey-m) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: @kbeaugrand, have you considered having the same implementation as the Azure SQL Server implementation?
See: https://github.com/microsoft/semantic-kernel/blob/e1ab3455c1c8ce40f434ab512cbb9093ba863c16/dotnet/src/Connectors/Connectors.Memory.SqlServer/SqlServerMemoryStore.cs#L57
It creates a new table for each collection using the collection name.  This means that no tables need to be created at startup time.  A table is only created when CreateCollection is called.

This is also good in that it means we have a consistent approach to modelling collections between the different SQL Server connector implementations.

[the remark this answers is not in the document]
Comment (kbeaugrand) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs]: I fixed it by removing this table and using only SK memories tables, by looking into tables with a prefix.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs [hunk 13]
```diff
@@ -0,0 +1,486 @@
...
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<bool> DoesCollectionExistsAsync(string collectionName,
+        CancellationToken cancellationToken = default)
+    {
+        collectionName = NormalizeIndexName(collectionName);
+
+        var collections = this.GetCollectionsAsync(cancellationToken)
```
Comment (westey-m) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/Core/SqlServerClient.cs:88]: This could be optimized by using a select that does `$"table_name = {this._configuration.MemoryTableName}_{collectionName}"` for if there is a large number of collections.
Can be in a separate PR though.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerConfig.cs [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+{
+    /// <summary>
+    /// The default SQL Server memories table name.
+    /// </summary>
+    internal const string DefaultMemoryTableName = "SKMemories";
+
+    /// <summary>
+    /// The default SQL Server embeddings table name.
+    /// </summary>
+    internal const string DefaultEmbeddingsTableName = "SKEmbeddings";
```
Comment (westey-m) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerConfig.cs:18]: We should update these names and descriptions to match the new behavior.
Consider MemoryTablePrefix or BaseMemoryTableName.
It'll be good to explain in the summary how these names would be used as prefixes for the tables when a collection is created.

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerConfig.cs [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+
+    /// <summary>
+    /// The SQL Server memories table name.
+    /// </summary>
+    public string MemoryTableName { get; set; } = DefaultMemoryTableName;
+
+    /// <summary>
+    /// The SQL Server embeddings table name.
+    /// </summary>
+    public string EmbeddingsTableName { get; set; } = DefaultEmbeddingsTableName;
```
Comment (westey-m) [inline dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerConfig.cs:43]: We should update these names and descriptions to match the new behavior.
Consider e.g. MemoryTablePrefix or BaseMemoryTableName.
It'll be good to explain in the summary how these names would be used as prefixes for the tables when a collection is created.

File: dotnet/src/IntegrationTests/Connectors/Memory/SqlServer/ClassicSqlServerMemoryStoreTests.cs [hunk 16]
```diff
@@ -0,0 +1,706 @@
...
+namespace SemanticKernel.IntegrationTests.Connectors.SqlServer;
+
+[System.Diagnostics.CodeAnalysis.SuppressMessage("Security", "CA2100:Review SQL queries for security vulnerabilities",
+                                                   Justification = "We need to build the full table name using schema and collection, it does not support parameterized passing.")]
+public class ClassicSqlServerMemoryStoreTests : IAsyncLifetime
+{
+    // If null, all tests will be enabled
+    //private const string? SkipReason = "Requires SqlServer server up and running";
+
+    private const string? SkipReason = null;
```
Comment (westey-m) [inline dotnet/src/IntegrationTests/Connectors/Memory/SqlServer/ClassicSqlServerMemoryStoreTests.cs:26]: Remember to swap this before check-in.

**Review by westey-m — APPROVED** [review]

**Review by westey-m — APPROVED** [review]

**Comment by markwallace-microsoft** [comment]
@kbeaugrand Can you run `dotnet format` and commit the resulting changes

File: dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs [hunk 17]
```diff
@@ -0,0 +1,246 @@
...
+        : this(new SqlServerClient(new SqlConnection(connectionString), config ?? new()))
+    {
+
+    }
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="SqlServerMemoryStore"/> class.
+    /// </summary>
+    /// <param name="dbClient">The SQL db client.</param>
+    internal SqlServerMemoryStore(SqlServerClient dbClient)
```
[the remark this answers is not in the document]
Comment (kbeaugrand) [reply dotnet/src/Connectors/Connectors.Memory.SqlServer.Classic/SqlServerMemoryStore.cs]: resolved

**Comment by kbeaugrand** [comment]
> @kbeaugrand Can you run `dotnet format` and commit the resulting changes

yes done !

**Comment by yorek** [comment]
@RogerBarreto @cincuranet just checking if we're ready to release this PR. Looking forward to blog about this

**Comment by cincuranet** [comment]
I think we are (minus the failures from CI).

**Comment by markwallace-microsoft** [comment]
@kbeaugrand Apologies for the delay in getting back to you. Can you resolve the merge conflict and run dotnet format. I'll do a second review early next week with the plan to get this merged soonest.

**Comment by markwallace-microsoft** [comment]
@kbeaugrand we recently announced support for the new [Microsoft.Extensions.VectorData.Abstractions](https://devblogs.microsoft.com/semantic-kernel/microsoft-extensions-vectordata-abstractions-now-available/). Our plan is to obsolete the old memory store abstractions so we're closing PR's that use those. If you are interested in updating to use the new Microsoft.Extensions.VectorData.Abstractions, we can consider this for additional the Semantic Kernel.
